src/webui/api/torrentscontroller.cpp

```diff
@@ -1041,3 +1041,112 @@ void TorrentsController::categoriesAction()
 
     setResult(categories);
 }
+
+void TorrentsController::addTagsAction()
+{
+    checkParams({"hashes", "tags"});
+
+    const QStringList hashes {params()["hashes"].split('|')};
+    const QStringList tags {params()["tags"].split(',', QString::SkipEmptyParts)};
+    QJsonArray errors;
+
+    for (const QString &tag : tags) {
+        const QString tagTrimmed {tag.trimmed()};
+        applyToTorrents(hashes, [&tagTrimmed, &errors](BitTorrent::TorrentHandle *const torrent)
+        {
+            if (!torrent->addTag(tagTrimmed) && !torrent->hasTag(tagTrimmed)) {
+                errors.append(QJsonObject {
+                    {"tag", tagTrimmed},
+                    {"hash", static_cast<QString>(torrent->hash())},
+                    {"errorMessage", "Unable to add tag"}
+                });
+            }
+        });
+    }
+
+    setResult(errors);
+}
+
+void TorrentsController::removeTagsAction()
+{
+    checkParams({"hashes"});
+
+    const QStringList hashes {params()["hashes"].split('|')};
+    const QStringList tags {params()["tags"].split(',', QString::SkipEmptyParts)};
+    QJsonArray errors;
+
+    for (const QString &tag : tags) {
+        const QString tagTrimmed {tag.trimmed()};
+        applyToTorrents(hashes, [&tagTrimmed, &errors](BitTorrent::TorrentHandle *const torrent)
+        {
+            if (!torrent->removeTag(tagTrimmed) && torrent->hasTag(tagTrimmed)) {
+                errors.append(QJsonObject {
+                    {"tag", tagTrimmed},
+                    {"hash", static_cast<QString>(torrent->hash())},
+                    {"errorMessage", "Unable to remove tag"}
+                });
+            }
+        });
+    }
+
+    if (tags.isEmpty()) {
+        applyToTorrents(hashes, [&errors](BitTorrent::TorrentHandle *const torrent)
+        {
+            torrent->removeAllTags();
+            if (!torrent->tags().isEmpty()) {
+                errors.append(QJsonObject {
+                    {"hash", static_cast<QString>(torrent->hash())},
+                    {"errorMessage", "Unable to remove tags"}
+                });
+            }
+        });
+    }
+
+    setResult(errors);
+}
+
+void TorrentsController::createTagsAction()
+{
+    checkParams({"tags"});
+
+    const QStringList tags {params()["tags"].split(',', QString::SkipEmptyParts)};
+    QJsonArray errors;
+
+    for (const QString &tag : tags) {
+        const QString tagTrimmed {tag.trimmed()};
+        if (!BitTorrent::Session::instance()->addTag(tagTrimmed) && !BitTorrent::Session::instance()->hasTag(tagTrimmed)) {
```

Comment:
I would change the order of conditionals.
Since C++ performs "Short circuit evaluation" you will not try to create tag if it already exists.

Applied in all similar cases.